### PR TITLE
Added graceful copy of TB_SceneOpened.js to TOONBOOM_GLOBAL_SCRIPT_LOCATION if this env exists

### PR DESCRIPTION
**Pull request by karmakat**
Added logic for graceful copy of `TB_sceneOpened.js` from **avalon-core** to `TOONBOOM_GLOBAL_SCRIPT_LOCATION` if the env already exists and the file is different.

- The previous logic would clobber the existing `TOONBOOM_GLOBAL_SCRIPT_LOCATION` , and therefore any custom scripts a studio had in that location would not be available on launch. 

- I suggest putting very clearly the fact that `TB_sceneOpened.js` will be overwritten in the `TOONBOOM_GLOBAL_SCRIPT_LOCATION` in documentation, and advising that TDs put their studio's Harmony startup logic in:
    - In a Harmony Package
    - In `TB_sceneOpenedUI.js` instead
    - In their own fork of **avalon-core**, adding to the `TB_sceneOpened.js`

### CODE REVIEW

**Comment by karmakat** [comment]
I can reformat this for the Hound violations, but since there can be a variety of os system errors when copying a file, I have left the exception general. Let me know what you'd like me to do.

**Comment by tokejepsen** [comment]
> I can reformat this for the Hound violations, but since there can be a variety of os system errors when copying a file, I have left the exception general. Let me know what you'd like me to do.

Yeah, could you fix the hound violations.

As for the exception, the best thing would be to have `except Exception as e` and log this error, along side the error logging you have currently, so at least we as developer can see it when debugging.

**Comment by mkolar** [comment]
We'll try this out, but apart from the hound, I see no problem with this.

File: avalon/harmony/lib.py [hunk 1]
```diff
@@ -55,7 +56,33 @@ def launch(application_path):
     self.application_path = application_path
 
     # Launch Harmony.
-    os.environ["TOONBOOM_GLOBAL_SCRIPT_LOCATION"] = os.path.dirname(__file__)
+    if os.getenv("TOONBOOM_GLOBAL_SCRIPT_LOCATION"):
+
+        avalon_harmony_startup_path = \
+            os.path.join(os.getenv("TOONBOOM_GLOBAL_SCRIPT_LOCATION"),
+                         "TB_sceneOpened.js")
```
Comment (tokejepsen) [inline avalon/harmony/lib.py:63]: This is more readable:

```
avalon_harmony_startup_path = os.path.join(
    os.getenv("TOONBOOM_GLOBAL_SCRIPT_LOCATION"), "TB_sceneOpened.js"
)
```

File: avalon/harmony/lib.py [hunk 2]
```diff
@@ -55,7 +56,33 @@ def launch(application_path):
...
+
+        env_harmony_startup_path = \
+            os.path.join(os.path.dirname(__file__), "TB_sceneOpened.js")
+
```
Comment (tokejepsen) [inline avalon/harmony/lib.py:67]:
```python
env_harmony_startup_path = os.path.join(
    os.path.dirname(__file__), "TB_sceneOpened.js"
)
```

File: avalon/harmony/lib.py [hunk 3]
```diff
@@ -55,7 +56,33 @@ def launch(application_path):
...
+        if not filecmp.cmp(avalon_harmony_startup_path,
```
Comment (tokejepsen) [inline avalon/harmony/lib.py:68]: What is `filecmp`?

Comment (karmakat) [reply avalon/harmony/lib.py:68]: Simple non-expensive byte comparison of 2 files. 
See: https://stackoverflow.com/questions/1072569/see-if-two-files-have-the-same-content-in-python

File: avalon/harmony/lib.py (hunk 1 above)
Comment (karmakat) [reply avalon/harmony/lib.py]: Done.

File: avalon/harmony/lib.py (hunk 2 above)
Comment (karmakat) [reply avalon/harmony/lib.py:68]: Done

File: avalon/harmony/lib.py (hunk 3 above)
Comment (tokejepsen) [reply avalon/harmony/lib.py:69]: Dont see any import?

Comment (tokejepsen) [reply avalon/harmony/lib.py:69]: Sorry, ignore that

File: avalon/harmony/lib.py [hunk 4]
```diff
@@ -55,7 +56,34 @@ def launch(application_path):
...
+            except Exception as e:
+                self.log.error(e)
+                self.log.warning(
+                    "Failed to copy {0} to {1}! "
+                    "Defaulting to Avalon TOONBOOM_GLOBAL_SCRIPT_LOCATION."
+                        .format(avalon_harmony_startup_path,
+                                env_harmony_startup_path))
+
+                os.environ["TOONBOOM_GLOBAL_SCRIPT_LOCATION"] = \
+                    os.path.dirname(__file__)
```
Comment (tokejepsen) [inline avalon/harmony/lib.py:83]: Sorry, I'm not a fan of `\` to escape line length. Better to assign to a variable.
```
path = os.path.dirname(__file__)
os.environ["TOONBOOM_GLOBAL_SCRIPT_LOCATION"] = path
```

File: avalon/harmony/lib.py [hunk 5]
```diff
@@ -55,7 +56,34 @@ def launch(application_path):
...
+    else:
+        os.environ["TOONBOOM_GLOBAL_SCRIPT_LOCATION"] = \
+            os.path.dirname(__file__)
```
Comment (tokejepsen) [inline avalon/harmony/lib.py:86]: Same as above.

**Comment by karmakat** [comment]
Ok... I believe this is ready to go. Hound was pretty merciless with whitespace violations on empty lines... not sure what was happening there since there actually was none.

**Comment by karmakat** [comment]
@tokejepsen @mkolar 
This logic also needs to be added to the ToonBoom folder. Please standby.

**Comment by karmakat** [comment]
@tokejepsen @mkolar 
I have refactored the logic into a function, and also used it in the avalon/toonboom folder. Toke says this still WIP, but agrees this logic should be added.

**Comment by karmakat** [comment]
Ok, @tokejepsen I made those changes.
Also, I discovered a typo where in the thread for launch of avalon server, thread.daemon was miss-spelled. I believe this may be the source of some issues I have had with the server, but I will need to test some more. Either way, it was definitely a mistake in the code.

File: avalon/harmony/lib.py [hunk 6]
```diff
@@ -109,7 +109,7 @@ def launch_zip_file(filepath):
     # Launch Avalon server.
     self.server = Server(self.port)
     thread = threading.Thread(target=self.server.start)
-    thread.deamon = True
+    thread.daemon = True
```
Comment (tokejepsen) [inline avalon/harmony/lib.py:112]: This is not a mispelling; https://docs.python.org/3/library/threading.html#threading.Thread.daemon

Please revert this.

Comment (karmakat) [reply avalon/harmony/lib.py:112]: Are you sure? It was spelt deamon but as the doc says that you posted it should be daemon.

Comment (tokejepsen) [reply avalon/harmony/lib.py:112]: Sorry my bad.

**Comment by mkolar** [comment]
I've tested on our end all it looks good. @tokejepsen thanks a lot for the code review. Good job guys. I'm happy to merge

**Pull request merged**